shared/src/ledger/eth_bridge/storage/bridge_pool.rs

```diff
@@ -44,3 +58,837 @@ pub fn is_bridge_pool_key(key: &Key) -> bool {
+        }
+
+        if hashes.is_empty() {
+            Default::default()
+        } else {
+            hashes.remove(0)
+        }
+    }
+
+    /// Return the root as a [`struct@Hash`] type.
+    pub fn root(&self) -> Hash {
+        self.root.clone().into()
+    }
+
+    /// Get a reference to the backing store
+    pub fn store(&self) -> &BTreeSet<KeccakHash> {
+        &self.store
+    }
+
+    /// Create a batched membership proof for the provided keys
+    pub fn get_membership_proof(
+        &self,
+        keys: &[Key],
+        mut values: Vec<PendingTransfer>,
+    ) -> Result<BridgePoolProof, Error> {
+        if values.len() != keys.len() {
+            return Err(eyre!(
+                "The number of leaves and leaf hashes must be equal."
+            )
+            .into());
+        }
+        // sort the values according to their hash values
+        values.sort_by_key(|transfer| transfer.keccak256());
+
+        // get the leaf hashes
+        let mut leaves: BTreeSet<KeccakHash> = Default::default();
+        for (key, value) in keys.iter().zip(values.iter()) {
+            let hash = Self::parse_key(key)?;
+            if hash != value.keccak256() {
+                return Err(eyre!(
+                    "Hashes of keys did not match hashes of values."
+                )
+                .into());
+            }
+            leaves.insert(hash);
+        }
+
+        let mut proof_hashes = vec![];
+        let mut flags = vec![];
+        let mut hashes: Vec<_> = self
+            .store
+            .iter()
+            .cloned()
+            .map(|hash| {
+                if leaves.contains(&hash) {
+                    Node::OnPath(hash)
+                } else {
+                    Node::Sibling(hash)
+                }
+            })
+            .collect();
+
+        while hashes.len() > 1 {
+            let mut next_hashes = vec![];
+
+            for pair in hashes.chunks(2) {
+                let left = pair[0].clone();
+                let right = pair.get(1).cloned().unwrap_or_default();
+                match (left, right) {
+                    (Node::OnPath(left), Node::OnPath(right)) => {
+                        flags.push(true);
+                        next_hashes
+                            .push(Node::OnPath(hash_pair(left.clone(), right)));
+                    }
+                    (Node::OnPath(hash), Node::Sibling(sib)) => {
+                        flags.push(false);
+                        proof_hashes.push(sib.clone());
+                        next_hashes
+                            .push(Node::OnPath(hash_pair(hash.clone(), sib)));
+                    }
+                    (Node::Sibling(sib), Node::OnPath(hash)) => {
+                        flags.push(false);
+                        proof_hashes.push(sib.clone());
+                        next_hashes
+                            .push(Node::OnPath(hash_pair(hash, sib.clone())));
+                    }
```

Comment:
We could combine these two match arms, something like:

```rust
                    (Node::OnPath(hash), Node::Sibling(sib))
                    | (Node::Sibling(sib), Node::OnPath(hash)) => {
                        flags.push(false);
                        proof_hashes.push(sib.clone());
                        next_hashes.push(Node::OnPath(hash_pair(
                            hash.to_owned(),
                            sib.to_owned(),
                        )));
                    }
```

Reply:
I tried that at first, but rust wouldn' have it. :weary: